lib/internal/errors.js

```diff
@@ -181,9 +176,56 @@ function makeNodeErrorWithCode(Base, key) {
         writable: true
       });
     }
+
+    toString() {
+      return `${this.name} [${key}]: ${this.message}`;
+    }
   };
 }
 
+// This function removes unnecessary frames from Node.js core errors.
+function hideStackFrames(fn) {
+  return function hidden(...args) {
+    // Make sure the most outer `hideStackFrames()` function is used.
+    let setStackFn = false;
+    if (excludedStackFn === undefined) {
+      excludedStackFn = hidden;
+      setStackFn = true;
+    }
+    try {
+      return fn(...args);
+    } finally {
+      if (setStackFn === true) {
+        excludedStackFn = undefined;
+      }
+    }
+  };
+}
+
+function addCodeToName(err, name, code) {
+  // Set the stack
+  if (excludedStackFn !== undefined) {
+    // eslint-disable-next-line no-restricted-syntax
+    Error.captureStackTrace(err, excludedStackFn);
+  }
+  // Add the error code to the name to include it in the stack trace.
+  err.name = `${name} [${code}]`;
+  // Access the stack to generate the error message including the error code
+  // from the name.
+  err.stack;
+  // Reset the name to the actual name.
+  if (name === 'SystemError') {
+    defineProperty(err, 'name', {
+      value: name,
+      enumerable: false,
+      writable: true,
+      configurable: true
+    });
+  } else {
+    delete err.name;
```

Comment:
Why do we need this? Can't we just reset `err.name = name`?

Reply:
Errors use their prototypes `name` and should not have a instances name. I add the instances name only to create the error stack. Removing it afterwards will fall back to the prototypes error name again.
`SystemError` works a bit different and that's why I set the name as non-enumerable property. I plan on refactoring that at some point but that seemed out of scope for this PR.